### DISCUSSION TITLE
Reproduce Results of Water dataset

### DISCUSSION THREAD

**Discussion by dariagusew** [mir-group/nequip#169] [Q&A] [answered]
Hi!
We wanted to reproduce your results with the dataset (https://github.com/BingqingCheng/ab-initio-thermodynamics-of-water) and have encountered a problem, while training the data. Our errors (only validation) start increasing after a couple of epochs, so we are probably in a way maybe overfitting the data ? I was wondering if you maybe encountered such a problem or if this has something to do with the hyperparameters ?
I will attach an excerpt of our training and also our configuration file. 
Thank you!

[water_training.txt](https://github.com/mir-group/nequip/files/8143177/water_training.txt)
[water_config.txt](https://github.com/mir-group/nequip/files/8143180/water_config.txt)

**Comment by simonbatzner** [comment, marked as answer]
Hi @dariagusew, I think I see the issue, explained below. On another note, please be sure to update to v0.5.3. of the code and to look at `f_mae` for your forces.

Regarding your issue: we have updated the default loss definition between v0.3.3 (used in the last arxiv version) and v0.5.3 (current version). In the loss definition, we use a joint loss that sums the force-loss and the energy-loss. In v0.3.3 we simply use the total energy loss and the force loss, weight each and sum them. In v0.5.3, our default is to use the the `PerAtomsMSELoss`, which normalizes the energy loss to a per-atom quantity. This is done since forces are a local quantity while the energy is a global quantity. So what `PerAtomsMSELoss` does is it uses a local per-atom energy loss by diving the global energy loss by N^2, where N comes from the fact that there are N atoms in the system (converting global to local) and the square deals with the fact that we use a MSE loss.

In version 0.3.3, we still used the total energy (i.e. not divided by the number of atoms N) and therefore had different weighting between the two. So in the paper we reported a weighting of 1 and 192^2 = 36864. So if you want to use the same config as I did in the paper you'd have to change this part in your config

```
loss_coeffs:                                                                       
  forces: 36846                                                                       
  total_energy:                                                                    
    - 1
    - PerAtomMSELoss
```

to this:

```
loss_coeffs:                                                                       
  forces: 36846                                                                       
  total_energy:  1
```

This will then *not* normalize the energy to a per-atom energy and should give you the same loss as what I used in the paper. What you did in your config is you combined PerAtomMSELoss with a high weight on the forces, that is equivalent to setting a force term of N^2 * 36,864 which gives something like 1.3 billion or so, so wildly off from what was used in the paper. In general, this relative weighting is a hyperparameter that is a function of your system. We typically recommend what we have in our example.yaml which is:

```
loss_coeffs:                                                                   
  forces: 1                                                                       
  total_energy:                                                              
    - 1
    - PerAtomMSELoss
```

This will do a 1:1 weighting, but with normalized energies, so in the language of the non-normalized weighting it will be force-weight N^2, energy-weight 1 for the reasons outlined above.

Please also be sure to use these metrics: 

```
# output metrics
metrics_components:
  - - forces                               # key 
    - mae                                  # "rmse" or "mae"
  - - forces
    - rmse
  - - forces
    - mae
    - PerSpecies: True                     # if true, per species contribution is counted separately
      report_per_component: False          # if true, statistics on each component (i.e. fx, fy, fz) will be counted separately
  - - forces                                
    - rmse                                  
    - PerSpecies: True                     
      report_per_component: False    
  - - total_energy
    - mae    
  - - total_energy
    - mae
    - PerAtom: True                        # if true, energy is normalized by the number of atoms
```

Obviously, for perfect reproducibility, we'd ask you to check out `v0.3.3` at the git commit specified in the paper and use the exact settings, but if you simply want a model that works as well, then the best thing to do is to run it in the current version, which includes a few improvements we've made since then and make the updates specified above.

**Reply by dariagusew** [reply]
Hi, Thank you so much for the help!
It works fine now!

**Reply by simonbatzner** [reply]
Great to hear :)
